File: tmp/patch-backend-nomina-block.mjs
import fs from 'node:fs';

function replaceOrThrow(source, searchValue, replaceValue, label) {
  const next = source.replace(searchValue, replaceValue);
  if (next === source) {
    throw new Error(`Replacement failed: ${label}`);
  }
  return next;
}

const servicePath = 'src/modules/empresa-configuracion/empresa-configuracion.service.ts';
let service = fs.readFileSync(servicePath, 'utf8');

service = replaceOrThrow(
  service,
  `export interface SalaryCategoryAssignmentApplyInput {\n  periodo_id: number;\n  target_category_id?: number | null;\n  nomina_empleado_ids: number[];\n  observacion?: string | null;\n}\n`,
  `export interface SalaryCategoryAssignmentCriteriaSnapshot\n  extends Omit<SalaryCategoryAssignmentPreviewInput, 'limit'> {}\n\nexport interface SalaryCategoryAssignmentApplyInput {\n  periodo_id: number;\n  target_category_id?: number | null;\n  nomina_empleado_ids: number[];\n  observacion?: string | null;\n  preview_criteria?: SalaryCategoryAssignmentCriteriaSnapshot | null;\n}\n`,
  'service apply input'
);

service = replaceOrThrow(
  service,
  `interface AssignmentCurrentRow extends QueryResultRow {\n  categoria_codigo: string | null;\n  categoria_id: string | null;\n  categoria_nombre: string | null;\n  nomina_empleado_id: string;\n  numero_documento: string | null;\n  persona_nombre: string;\n}\n`,
  `interface AssignmentCurrentRow extends QueryResultRow {\n  categoria_codigo: string | null;\n  categoria_id: string | null;\n  categoria_nombre: string | null;\n  nomina_empleado_id: string;\n  numero_documento: string | null;\n  persona_nombre: string;\n}\n\ninterface AssignmentModalidadOptionRow extends QueryResultRow {\n  modalidad: string | null;\n  modalidad_codigo: string | null;\n  total_trabajadores: number;\n}\n`,
  'service options row'
);

service = replaceOrThrow(
  service,
  'const buildAssignmentPreviewQuery = (',
  'export const buildAssignmentPreviewQuery = (',
  'service export buildAssignmentPreviewQuery'
);

service = replaceOrThrow(
  service,
  `const loadTargetCategoryForPreview = async (\n`,
  `const intersectsDateRange = (\n  rangeAStart: string | null | undefined,\n  rangeAEnd: string | null | undefined,\n  rangeBStart: string | null | undefined,\n  rangeBEnd: string | null | undefined\n) => {\n  if (!rangeAStart || !rangeBStart) {\n    return false;\n  }\n\n  const normalizedRangeAEnd = rangeAEnd ?? '9999-12-31';\n  const normalizedRangeBEnd = rangeBEnd ?? '9999-12-31';\n  return normalizedRangeAEnd >= rangeBStart && normalizedRangeBEnd >= rangeAStart;\n};\n\nexport const buildSalaryCategoryAssignmentCriteriaSnapshot = (\n  input: SalaryCategoryAssignmentCriteriaSnapshot | null | undefined\n): SalaryCategoryAssignmentCriteriaSnapshot | null => {\n  if (!input) {\n    return null;\n  }\n\n  return Object.fromEntries(\n    Object.entries({\n      periodo_id: input.periodo_id,\n      target_category_id: input.target_category_id ?? null,\n      search: trimNullable(input.search),\n      contrato_cargo_id: input.contrato_cargo_id ?? undefined,\n      cargo: trimNullable(input.cargo),\n      municipio_id: input.municipio_id ?? undefined,\n      municipio: trimNullable(input.municipio),\n      institucion_id: input.institucion_id ?? undefined,\n      institucion: trimNullable(input.institucion),\n      sede_id: input.sede_id ?? undefined,\n      sede: trimNullable(input.sede),\n      modalidad_id: input.modalidad_id ?? undefined,\n      modalidad: trimNullable(input.modalidad),\n      modalidad_codigo: trimNullable(input.modalidad_codigo)?.toUpperCase() ?? undefined,\n      metodo_pago: trimNullable(input.metodo_pago)?.toUpperCase() ?? undefined,\n      estado_vinculacion: trimNullable(input.estado_vinculacion)?.toUpperCase() ?? undefined,\n      vinculacion_activa: input.vinculacion_activa ?? undefined,\n      institucion_sede_count: input.institucion_sede_count\n        ? {\n            operator: input.institucion_sede_count.operator,\n            value: input.institucion_sede_count.value ?? undefined,\n            min: input.institucion_sede_count.min ?? undefined,\n            max: input.institucion_sede_count.max ?? undefined\n          }\n        : undefined,\n      without_category: input.without_category === true ? true : undefined\n    }).filter(([, value]) => value !== undefined)\n  ) as SalaryCategoryAssignmentCriteriaSnapshot;\n};\n\nconst resolveSalaryCategoryAssignmentAction = (\n  currentCategoryId: string | null,\n  nextCategoryId: number | null\n): 'ASIGNAR' | 'CAMBIAR' | 'RETIRAR' => {\n  if (!currentCategoryId && nextCategoryId) {\n    return 'ASIGNAR';\n  }\n\n  if (currentCategoryId && !nextCategoryId) {\n    return 'RETIRAR';\n  }\n\n  return 'CAMBIAR';\n};\n\nexport const listSalaryCategoryAssignmentOptions = async (\n  empresaId: number,\n  periodoId: number,\n  tenant?: TenantAccessContext\n) => {\n  await loadPeriodoScopeOrThrow(periodoId, empresaId, tenant);\n\n  const result = await dbQuery<AssignmentModalidadOptionRow>(\n    `SELECT\n        modalidad_codigo,\n        modalidad,\n        COUNT(*)::int AS total_trabajadores\n      FROM (\n        ${buildAssignmentBaseSql()}\n      ) assignment_source\n      WHERE COALESCE(modalidad_codigo, '') <> ''\n      GROUP BY modalidad_codigo, modalidad\n      ORDER BY modalidad_codigo ASC, modalidad ASC`,\n    [periodoId]\n  );\n\n  return {\n    modalidades: result.rows.map((row) => ({\n      modalidad_codigo: row.modalidad_codigo ?? '',\n      modalidad: row.modalidad,\n      total_trabajadores: Number(row.total_trabajadores ?? 0)\n    }))\n  };\n};\n\nconst loadTargetCategoryForPreview = async (\n`,
  'service helpers insertion'
);

service = replaceOrThrow(
  service,
  `  const row = await loadSalaryCategoryRowOrThrow(targetCategoryId, tenant, client);\n  if (Number(row.contrato_id) !== Number(periodo.contrato_id)) {\n    throw new AppError(\n      'Salary category does not belong to the payroll period contract',\n      409,\n      'CATEGORY_PERIOD_CONTRACT_MISMATCH'\n    );\n  }\n\n  await assertContractBelongsToCompany(Number(row.contrato_id), empresaId);\n  return mapSalaryCategory(row);\n};\n`,
  `  const row = await loadSalaryCategoryRowOrThrow(targetCategoryId, tenant, client);\n  if (Number(row.contrato_id) !== Number(periodo.contrato_id)) {\n    throw new AppError(\n      'Salary category does not belong to the payroll period contract',\n      409,\n      'CATEGORY_PERIOD_CONTRACT_MISMATCH'\n    );\n  }\n\n  await assertContractBelongsToCompany(Number(row.contrato_id), empresaId);\n\n  const category = mapSalaryCategory(row);\n  if (!category.activo) {\n    throw new AppError('Salary category is inactive', 409, 'CATEGORY_INACTIVE');\n  }\n\n  if (\n    !intersectsDateRange(\n      category.vigente_desde,\n      category.vigente_hasta ?? null,\n      toDateString(periodo.fecha_inicio),\n      toDateString(periodo.fecha_fin)\n    )\n  ) {\n    throw new AppError(\n      'Salary category is outside the payroll period validity',\n      409,\n      'CATEGORY_OUT_OF_PERIOD_RANGE'\n    );\n  }\n\n  return category;\n};\n`,
  'service category validation'
);

service = replaceOrThrow(
  service,
  `    const category = await loadTargetCategoryForPreview(\n      empresaId,\n      periodo,\n      input.target_category_id,\n      tenant,\n      client\n    );\n    const targetCategoryId = category ? Number(category.id) : null;\n\n    const employeesResult = await client.query<AssignmentCurrentRow>(\n`,
  `    const category = await loadTargetCategoryForPreview(\n      empresaId,\n      periodo,\n      input.target_category_id,\n      tenant,\n      client\n    );\n    const targetCategoryId = category ? Number(category.id) : null;\n    const criteriaSnapshot = buildSalaryCategoryAssignmentCriteriaSnapshot(\n      input.preview_criteria ?? {\n        periodo_id: input.periodo_id,\n        target_category_id: input.target_category_id ?? null\n      }\n    );\n\n    const employeesResult = await client.query<AssignmentCurrentRow>(\n`,
  'service criteria snapshot apply'
);

service = replaceOrThrow(
  service,
  `    for (const row of changedRows) {\n      const before = {\n        nomina_empleado_id: row.nomina_empleado_id,\n        categoria_salarial_id: row.categoria_id,\n        categoria_salarial: row.categoria_id\n          ? {\n              id: row.categoria_id,\n              codigo_categoria: row.categoria_codigo,\n              nombre_categoria: row.categoria_nombre\n            }\n          : null\n      };\n\n      await client.query(\n`,
  `    for (const row of changedRows) {\n      const assignmentAction = resolveSalaryCategoryAssignmentAction(row.categoria_id, targetCategoryId);\n      const before = {\n        nomina_empleado_id: row.nomina_empleado_id,\n        categoria_salarial_id: row.categoria_id,\n        categoria_salarial: row.categoria_id\n          ? {\n              id: row.categoria_id,\n              codigo_categoria: row.categoria_codigo,\n              nombre_categoria: row.categoria_nombre\n            }\n          : null\n      };\n\n      await client.query(\n`,
  'service assignment action'
);

service = replaceOrThrow(
  service,
  `      await registerAuditEntry({\n        client,\n        usuario_id: actorUserId,\n        accion: 'NOMINA_EMPLEADO_CATEGORY_ASSIGN',\n        tabla: 'nomina_empleados',\n        registro_id: row.nomina_empleado_id,\n        descripcion: input.observacion ?? 'Asignación manual de categoría salarial',\n        contrato_id: String(periodo.contrato_id),\n        empresa_id: String(periodo.empresa_id),\n        before: {\n          ...before,\n          periodo_id: String(periodo.id),\n          contrato_id: String(periodo.contrato_id),\n          empresa_id: String(periodo.empresa_id)\n        },\n        after: {\n          nomina_empleado_id: row.nomina_empleado_id,\n          periodo_id: String(periodo.id),\n          contrato_id: String(periodo.contrato_id),\n          empresa_id: String(periodo.empresa_id),\n          categoria_salarial_id: targetCategoryId ? String(targetCategoryId) : null,\n          categoria_salarial: category\n            ? {\n                id: category.id,\n                codigo_categoria: category.codigo_categoria,\n                nombre_categoria: category.nombre_categoria\n              }\n            : null,\n          trabajador: {\n            nombre_completo: row.persona_nombre,\n            numero_documento: row.numero_documento\n          }\n        },\n        ip: auditMeta?.ip ?? null,\n        user_agent: auditMeta?.user_agent ?? null\n      });\n`,
  `      await registerAuditEntry({\n        client,\n        usuario_id: actorUserId,\n        accion: 'NOMINA_EMPLEADO_CATEGORY_ASSIGN',\n        tabla: 'nomina_empleados',\n        registro_id: row.nomina_empleado_id,\n        descripcion: input.observacion ?? `${assignmentAction} manual de categoría salarial`,\n        contrato_id: String(periodo.contrato_id),\n        empresa_id: String(periodo.empresa_id),\n        before: {\n          ...before,\n          periodo_id: String(periodo.id),\n          contrato_id: String(periodo.contrato_id),\n          empresa_id: String(periodo.empresa_id),\n          accion_categoria: assignmentAction,\n          observacion: input.observacion ?? null,\n          criterios_preview: criteriaSnapshot,\n          trabajador: {\n            nombre_completo: row.persona_nombre,\n            numero_documento: row.numero_documento\n          }\n        },\n        after: {\n          nomina_empleado_id: row.nomina_empleado_id,\n          periodo_id: String(periodo.id),\n          contrato_id: String(periodo.contrato_id),\n          empresa_id: String(periodo.empresa_id),\n          accion_categoria: assignmentAction,\n          observacion: input.observacion ?? null,\n          criterios_preview: criteriaSnapshot,\n          categoria_salarial_id: targetCategoryId ? String(targetCategoryId) : null,\n          categoria_salarial: category\n            ? {\n                id: category.id,\n                codigo_categoria: category.codigo_categoria,\n                nombre_categoria: category.nombre_categoria\n              }\n            : null,\n          trabajador: {\n            nombre_completo: row.persona_nombre,\n            numero_documento: row.numero_documento\n          }\n        },\n        ip: auditMeta?.ip ?? null,\n        user_agent: auditMeta?.user_agent ?? null\n      });\n`,
  'service audit enrichment'
);

fs.writeFileSync(servicePath, service);

const routesPath = 'src/modules/empresa-configuracion/empresa-configuracion.routes.ts';
let routes = fs.readFileSync(routesPath, 'utf8');

routes = replaceOrThrow(
  routes,
  `  listPayrollParameters,\n  listSalaryCategories,\n  previewSalaryCategoryAssignment,\n`,
  `  listPayrollParameters,\n  listSalaryCategories,\n  listSalaryCategoryAssignmentOptions,\n  previewSalaryCategoryAssignment,\n`,
  'routes import options'
);

routes = replaceOrThrow(
  routes,
  `const assignmentPreview = z.object({\n`,
  `const assignmentPreview = z.object({\n`,
  'routes anchor preview'
);

routes = replaceOrThrow(
  routes,
  `const assignmentApply = z.object({\n  periodo_id: z.coerce.number().int().positive(),\n  target_category_id: z.coerce.number().int().positive().nullable().optional(),\n  nomina_empleado_ids: z.array(z.coerce.number().int().positive()).min(1),\n  observacion: z.string().trim().max(500).nullable().optional()\n});\n`,
  `const assignmentOptionsQuery = z.object({\n  periodo_id: z.coerce.number().int().positive()\n});\nconst assignmentApply = z.object({\n  periodo_id: z.coerce.number().int().positive(),\n  target_category_id: z.coerce.number().int().positive().nullable().optional(),\n  nomina_empleado_ids: z.array(z.coerce.number().int().positive()).min(1),\n  observacion: z.string().trim().max(500).nullable().optional(),\n  preview_criteria: assignmentPreview.omit({ limit: true }).nullable().optional()\n});\n`,
  'routes apply schema'
);

routes = replaceOrThrow(
  routes,
  `router.post(\n  '/:empresaId/salary-categories/assignments/preview',\n`,
  `router.get(\n  '/:empresaId/salary-categories/assignments/options',\n  requirePermissions('nomina.economico.read'),\n  asyncHandler(async (req, res) =>\n    res.json({\n      data: await listSalaryCategoryAssignmentOptions(\n        id.parse(req.params.empresaId),\n        assignmentOptionsQuery.parse(req.query).periodo_id,\n        req.tenant\n      )\n    })\n  )\n);\n\nrouter.post(\n  '/:empresaId/salary-categories/assignments/preview',\n`,
  'routes options endpoint'
);

fs.writeFileSync(routesPath, routes);
console.log('patched backend files');

File: tmp/patch-backend-nomina-block.py
# -*- coding: utf-8 -*-
from pathlib import Path


def replace_or_throw(source: str, old: str, new: str, label: str) -> str:
    if old not in source:
        raise RuntimeError(f'Replacement failed: {label}')
    return source.replace(old, new, 1)

service_path = Path('src/modules/empresa-configuracion/empresa-configuracion.service.ts')
service = service_path.read_text(encoding='utf-8')

service = replace_or_throw(
    service,
    """export interface SalaryCategoryAssignmentApplyInput {
  periodo_id: number;
  target_category_id?: number | null;
  nomina_empleado_ids: number[];
  observacion?: string | null;
}
""",
    """export interface SalaryCategoryAssignmentCriteriaSnapshot
  extends Omit<SalaryCategoryAssignmentPreviewInput, 'limit'> {}

export interface SalaryCategoryAssignmentApplyInput {
  periodo_id: number;
  target_category_id?: number | null;
  nomina_empleado_ids: number[];
  observacion?: string | null;
  preview_criteria?: SalaryCategoryAssignmentCriteriaSnapshot | null;
}
""",
    'service apply input',
)

service = replace_or_throw(
    service,
    """interface AssignmentCurrentRow extends QueryResultRow {
  categoria_codigo: string | null;
  categoria_id: string | null;
  categoria_nombre: string | null;
  nomina_empleado_id: string;
  numero_documento: string | null;
  persona_nombre: string;
}
""",
    """interface AssignmentCurrentRow extends QueryResultRow {
  categoria_codigo: string | null;
  categoria_id: string | null;
  categoria_nombre: string | null;
  nomina_empleado_id: string;
  numero_documento: string | null;
  persona_nombre: string;
}

interface AssignmentModalidadOptionRow extends QueryResultRow {
  modalidad: string | null;
  modalidad_codigo: string | null;
  total_trabajadores: number;
}
""",
    'service options row',
)

service = replace_or_throw(
    service,
    'const buildAssignmentPreviewQuery = (',
    'export const buildAssignmentPreviewQuery = (',
    'service export buildAssignmentPreviewQuery',
)

helpers = """const intersectsDateRange = (
  rangeAStart: string | null | undefined,
  rangeAEnd: string | null | undefined,
  rangeBStart: string | null | undefined,
  rangeBEnd: string | null | undefined
) => {
  if (!rangeAStart || !rangeBStart) {
    return false;
  }

  const normalizedRangeAEnd = rangeAEnd ?? '9999-12-31';
  const normalizedRangeBEnd = rangeBEnd ?? '9999-12-31';
  return normalizedRangeAEnd >= rangeBStart && normalizedRangeBEnd >= rangeAStart;
};

export const buildSalaryCategoryAssignmentCriteriaSnapshot = (
  input: SalaryCategoryAssignmentCriteriaSnapshot | null | undefined
): SalaryCategoryAssignmentCriteriaSnapshot | null => {
  if (!input) {
    return null;
  }

  return Object.fromEntries(
    Object.entries({
      periodo_id: input.periodo_id,
      target_category_id: input.target_category_id ?? null,
      search: trimNullable(input.search),
      contrato_cargo_id: input.contrato_cargo_id ?? undefined,
      cargo: trimNullable(input.cargo),
      municipio_id: input.municipio_id ?? undefined,
      municipio: trimNullable(input.municipio),
      institucion_id: input.institucion_id ?? undefined,
      institucion: trimNullable(input.institucion),
      sede_id: input.sede_id ?? undefined,
      sede: trimNullable(input.sede),
      modalidad_id: input.modalidad_id ?? undefined,
      modalidad: trimNullable(input.modalidad),
      modalidad_codigo: trimNullable(input.modalidad_codigo)?.toUpperCase() ?? undefined,
      metodo_pago: trimNullable(input.metodo_pago)?.toUpperCase() ?? undefined,
      estado_vinculacion: trimNullable(input.estado_vinculacion)?.toUpperCase() ?? undefined,
      vinculacion_activa: input.vinculacion_activa ?? undefined,
      institucion_sede_count: input.institucion_sede_count
        ? {
            operator: input.institucion_sede_count.operator,
            value: input.institucion_sede_count.value ?? undefined,
            min: input.institucion_sede_count.min ?? undefined,
            max: input.institucion_sede_count.max ?? undefined
          }
        : undefined,
      without_category: input.without_category === true ? true : undefined
    }).filter(([, value]) => value !== undefined)
  ) as SalaryCategoryAssignmentCriteriaSnapshot;
};

const resolveSalaryCategoryAssignmentAction = (
  currentCategoryId: string | null,
  nextCategoryId: number | null
): 'ASIGNAR' | 'CAMBIAR' | 'RETIRAR' => {
  if (!currentCategoryId && nextCategoryId) {
    return 'ASIGNAR';
  }

  if (currentCategoryId && !nextCategoryId) {
    return 'RETIRAR';
  }

  return 'CAMBIAR';
};

export const listSalaryCategoryAssignmentOptions = async (
  empresaId: number,
  periodoId: number,
  tenant?: TenantAccessContext
) => {
  await loadPeriodoScopeOrThrow(periodoId, empresaId, tenant);

  const result = await dbQuery<AssignmentModalidadOptionRow>(
    `SELECT
        modalidad_codigo,
        modalidad,
        COUNT(*)::int AS total_trabajadores
      FROM (
        ${buildAssignmentBaseSql()}
      ) assignment_source
      WHERE COALESCE(modalidad_codigo, '') <> ''
      GROUP BY modalidad_codigo, modalidad
      ORDER BY modalidad_codigo ASC, modalidad ASC`,
    [periodoId]
  );

  return {
    modalidades: result.rows.map((row) => ({
      modalidad_codigo: row.modalidad_codigo ?? '',
      modalidad: row.modalidad,
      total_trabajadores: Number(row.total_trabajadores ?? 0)
    }))
  };
};

const loadTargetCategoryForPreview = async (
"""

service = replace_or_throw(
    service,
    "const loadTargetCategoryForPreview = async (\n",
    helpers,
    'service helpers insertion',
)

service = replace_or_throw(
    service,
    """  const row = await loadSalaryCategoryRowOrThrow(targetCategoryId, tenant, client);
  if (Number(row.contrato_id) !== Number(periodo.contrato_id)) {
    throw new AppError(
      'Salary category does not belong to the payroll period contract',
      409,
      'CATEGORY_PERIOD_CONTRACT_MISMATCH'
    );
  }

  await assertContractBelongsToCompany(Number(row.contrato_id), empresaId);
  return mapSalaryCategory(row);
};
""",
    """  const row = await loadSalaryCategoryRowOrThrow(targetCategoryId, tenant, client);
  if (Number(row.contrato_id) !== Number(periodo.contrato_id)) {
    throw new AppError(
      'Salary category does not belong to the payroll period contract',
      409,
      'CATEGORY_PERIOD_CONTRACT_MISMATCH'
    );
  }

  await assertContractBelongsToCompany(Number(row.contrato_id), empresaId);

  const category = mapSalaryCategory(row);
  if (!category.activo) {
    throw new AppError('Salary category is inactive', 409, 'CATEGORY_INACTIVE');
  }

  if (
    !intersectsDateRange(
      category.vigente_desde,
      category.vigente_hasta ?? null,
      toDateString(periodo.fecha_inicio),
      toDateString(periodo.fecha_fin)
    )
  ) {
    throw new AppError(
      'Salary category is outside the payroll period validity',
      409,
      'CATEGORY_OUT_OF_PERIOD_RANGE'
    );
  }

  return category;
};
""",
    'service category validation',
)

service = replace_or_throw(
    service,
    """    const category = await loadTargetCategoryForPreview(
      empresaId,
      periodo,
      input.target_category_id,
      tenant,
      client
    );
    const targetCategoryId = category ? Number(category.id) : null;

    const employeesResult = await client.query<AssignmentCurrentRow>(
""",
    """    const category = await loadTargetCategoryForPreview(
      empresaId,
      periodo,
      input.target_category_id,
      tenant,
      client
    );
    const targetCategoryId = category ? Number(category.id) : null;
    const criteriaSnapshot = buildSalaryCategoryAssignmentCriteriaSnapshot(
      input.preview_criteria ?? {
        periodo_id: input.periodo_id,
        target_category_id: input.target_category_id ?? null
      }
    );

    const employeesResult = await client.query<AssignmentCurrentRow>(
""",
    'service criteria snapshot apply',
)

service = replace_or_throw(
    service,
    """    for (const row of changedRows) {
      const before = {
        nomina_empleado_id: row.nomina_empleado_id,
        categoria_salarial_id: row.categoria_id,
        categoria_salarial: row.categoria_id
          ? {
              id: row.categoria_id,
              codigo_categoria: row.categoria_codigo,
              nombre_categoria: row.categoria_nombre
            }
          : null
      };

      await client.query(
""",
    """    for (const row of changedRows) {
      const assignmentAction = resolveSalaryCategoryAssignmentAction(row.categoria_id, targetCategoryId);
      const before = {
        nomina_empleado_id: row.nomina_empleado_id,
        categoria_salarial_id: row.categoria_id,
        categoria_salarial: row.categoria_id
          ? {
              id: row.categoria_id,
              codigo_categoria: row.categoria_codigo,
              nombre_categoria: row.categoria_nombre
            }
          : null
      };

      await client.query(
""",
    'service assignment action',
)

service = replace_or_throw(
    service,
    """      await registerAuditEntry({
        client,
        usuario_id: actorUserId,
        accion: 'NOMINA_EMPLEADO_CATEGORY_ASSIGN',
        tabla: 'nomina_empleados',
        registro_id: row.nomina_empleado_id,
        descripcion: input.observacion ?? 'Asignación manual de categoría salarial',
        contrato_id: String(periodo.contrato_id),
        empresa_id: String(periodo.empresa_id),
        before: {
          ...before,
          periodo_id: String(periodo.id),
          contrato_id: String(periodo.contrato_id),
          empresa_id: String(periodo.empresa_id)
        },
        after: {
          nomina_empleado_id: row.nomina_empleado_id,
          periodo_id: String(periodo.id),
          contrato_id: String(periodo.contrato_id),
          empresa_id: String(periodo.empresa_id),
          categoria_salarial_id: targetCategoryId ? String(targetCategoryId) : null,
          categoria_salarial: category
            ? {
                id: category.id,
                codigo_categoria: category.codigo_categoria,
                nombre_categoria: category.nombre_categoria
              }
            : null,
          trabajador: {
            nombre_completo: row.persona_nombre,
            numero_documento: row.numero_documento
          }
        },
        ip: auditMeta?.ip ?? null,
        user_agent: auditMeta?.user_agent ?? null
      });
""",
    """      await registerAuditEntry({
        client,
        usuario_id: actorUserId,
        accion: 'NOMINA_EMPLEADO_CATEGORY_ASSIGN',
        tabla: 'nomina_empleados',
        registro_id: row.nomina_empleado_id,
        descripcion: input.observacion ?? `${assignmentAction} manual de categoría salarial`,
        contrato_id: String(periodo.contrato_id),
        empresa_id: String(periodo.empresa_id),
        before: {
          ...before,
          periodo_id: String(periodo.id),
          contrato_id: String(periodo.contrato_id),
          empresa_id: String(periodo.empresa_id),
          accion_categoria: assignmentAction,
          observacion: input.observacion ?? null,
          criterios_preview: criteriaSnapshot,
          trabajador: {
            nombre_completo: row.persona_nombre,
            numero_documento: row.numero_documento
          }
        },
        after: {
          nomina_empleado_id: row.nomina_empleado_id,
          periodo_id: String(periodo.id),
          contrato_id: String(periodo.contrato_id),
          empresa_id: String(periodo.empresa_id),
          accion_categoria: assignmentAction,
          observacion: input.observacion ?? null,
          criterios_preview: criteriaSnapshot,
          categoria_salarial_id: targetCategoryId ? String(targetCategoryId) : null,
          categoria_salarial: category
            ? {
                id: category.id,
                codigo_categoria: category.codigo_categoria,
                nombre_categoria: category.nombre_categoria
              }
            : null,
          trabajador: {
            nombre_completo: row.persona_nombre,
            numero_documento: row.numero_documento
          }
        },
        ip: auditMeta?.ip ?? null,
        user_agent: auditMeta?.user_agent ?? null
      });
""",
    'service audit enrichment',
)

service_path.write_text(service, encoding='utf-8')

routes_path = Path('src/modules/empresa-configuracion/empresa-configuracion.routes.ts')
routes = routes_path.read_text(encoding='utf-8')

routes = replace_or_throw(
    routes,
    """  listPayrollParameters,
  listSalaryCategories,
  previewSalaryCategoryAssignment,
""",
    """  listPayrollParameters,
  listSalaryCategories,
  listSalaryCategoryAssignmentOptions,
  previewSalaryCategoryAssignment,
""",
    'routes import options',
)

routes = replace_or_throw(
    routes,
    """const assignmentApply = z.object({
  periodo_id: z.coerce.number().int().positive(),
  target_category_id: z.coerce.number().int().positive().nullable().optional(),
  nomina_empleado_ids: z.array(z.coerce.number().int().positive()).min(1),
  observacion: z.string().trim().max(500).nullable().optional()
});
""",
    """const assignmentOptionsQuery = z.object({
  periodo_id: z.coerce.number().int().positive()
});
const assignmentApply = z.object({
  periodo_id: z.coerce.number().int().positive(),
  target_category_id: z.coerce.number().int().positive().nullable().optional(),
  nomina_empleado_ids: z.array(z.coerce.number().int().positive()).min(1),
  observacion: z.string().trim().max(500).nullable().optional(),
  preview_criteria: assignmentPreview.omit({ limit: true }).nullable().optional()
});
""",
    'routes apply schema',
)

routes = replace_or_throw(
    routes,
    """router.post(
  '/:empresaId/salary-categories/assignments/preview',
""",
    """router.get(
  '/:empresaId/salary-categories/assignments/options',
  requirePermissions('nomina.economico.read'),
  asyncHandler(async (req, res) =>
    res.json({
      data: await listSalaryCategoryAssignmentOptions(
        id.parse(req.params.empresaId),
        assignmentOptionsQuery.parse(req.query).periodo_id,
        req.tenant
      )
    })
  )
);

router.post(
  '/:empresaId/salary-categories/assignments/preview',
""",
    'routes options endpoint',
)

routes_path.write_text(routes, encoding='utf-8')
print('patched backend files')

